### PR TITLE
Find optimal threshold on raw data

### PR DESCRIPTION
**Pull request by charleygros**
Two changes:
- Instead of doing the binarization threshold search on the transformed data (ie after Resampling / Cropping), we do it on the Raw data
- Instead of only considering the validation sub-dataset, we now use training+validation to have a better estimation of this threshold

### CODE REVIEW

File: ivadomed/testing.py [hunk 1]
```diff
@@ -243,11 +247,99 @@ def run_inference(test_loader, model, model_params, testing_params, ofolder, cud
...
+                        num_workers=0)
+
+    # Run inference
+    preds_npy, gt_npy = run_inference(loader, model, model_params,
+                                      testing_params,
+                                      ofolder=None,
+                                      cuda_available=cuda_available)
+
+    # Make sure the GT is binarized
+    gt_npy = [threshold_predictions(gt, thr=0.5) for gt in gt_npy]
```
Comment (andreanne-lemay) [inline ivadomed/testing.py:309]: I don't think a loop is necessary here

```suggestion
    gt_npy = threshold_predictions(gt_npy, thr=0.5)
```

File: ivadomed/testing.py [hunk 2]
```diff
@@ -243,11 +247,99 @@ def run_inference(test_loader, model, model_params, testing_params, ofolder, cud
...
+    # Move threshold
+    for thr in thr_list:
+        preds_thr = [threshold_predictions(copy.deepcopy(pred), thr=thr) for pred in preds_npy]
```
Comment (andreanne-lemay) [inline ivadomed/testing.py:312]: same comment

```suggestion
        preds_thr = threshold_predictions(copy.deepcopy(preds_npy), thr=thr)
```

**Review by andreanne-lemay — APPROVED** [review]
Looks good to me :100:

File: ivadomed/main.py [hunk 3]
```diff
@@ -189,8 +202,31 @@ def run_command(context, n_gif=0, thr_increment=None):
             thr_increment=thr_increment,
```
Comment (andreanne-lemay) [inline ivadomed/main.py:202]: This variable is no longer used
```suggestion
```

File: ivadomed/main.py [hunk 4]
```diff
@@ -189,8 +202,31 @@ def run_command(context, n_gif=0, thr_increment=None):
...
             debugging=context["debugging"])
 
-        # Update threshold in config file
         if thr_increment:
+            # Choice of optimisation metric
```
Comment (andreanne-lemay) [inline ivadomed/main.py:206]: What do you think about putting all this block directly in testing, this way we don't need to retrain to have an optimal threshold?

File: ivadomed/main.py [hunk 5]
```diff
@@ -210,6 +210,18 @@ def run_command(context, n_gif=0, thr_increment=None):
...
     if thr_increment:
+        # LOAD DATASET
+        # Get Validation dataset
+        ds_valid = imed_loader.load_dataset(**{**loader_params,
+                                               **{'data_list': valid_lst, 'transforms_params': transform_valid_params,
+                                                  'dataset_type': 'validation'}}, device=device,
+                                            cuda_available=cuda_available)
+        # Get Training dataset
+        ds_train = imed_loader.load_dataset(**{**loader_params,
+                                               **{'data_list': train_lst, 'transforms_params': transform_train_params,
```
Comment (charleygros) [inline ivadomed/main.py:220]: @andreanne-lemay --> needs transform_valid here, otherwise DA active! Will change this

**Pull request merged**